### ISSUE TITLE
can't import pysam 0.9.1.2 (OS X)

### ISSUE THREAD

**Issue by leekaiinthesky** [pysam-developers/pysam#308]
On OS X 10.11.5, installed with pip.

Sorry, I realize this is intimately related to many other recent issues (both open and closed), but I wasn't sure which one was most relevant, so I created a new issue since this relates specifically to 0.9.1.2 and the exact error message is new:

```
>>> import pysam
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python2.7/site-packages/pysam/__init__.py", line 5, in <module>
    from pysam.libchtslib import *
ImportError: dlopen(/usr/local/lib/python2.7/site-packages/pysam/libchtslib.so, 2): Symbol not found: _hfile_plugin_init_libcurl
  Referenced from: /usr/local/lib/python2.7/site-packages/pysam/libchtslib.so
  Expected in: flat namespace
 in /usr/local/lib/python2.7/site-packages/pysam/libchtslib.so
```

Reverting to a previous version of pysam seemed to work fine for me. Thanks!

**Comment by david4096** [comment]
Also saw the same thing on Mac 10.11.5 and python 2.7.12 (thanks @Arc39).

**Comment by AndreasHeger** [comment]
Thanks both. I have made a new release (0.9.1.3) which - hopefully - will resolve this problem. I am closing this issue, instead please see #306

**Comment by leekaiinthesky** [comment]
Doesn't seem to be quite the same. For us (OS X 10.11.5), installation via pip is successful; only import fails with `Symbol not found: _hfile_plugin_init_libcurl`.

Just checked 0.9.1.3 in both Python 2.7 and 3.4 and got the same error message as before:
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python3.4/site-packages/pysam/__init__.py", line 5, in <module>
    from pysam.libchtslib import *
ImportError: dlopen(/usr/local/lib/python3.4/site-packages/pysam/libchtslib.so, 2): Symbol not found: _hfile_plugin_init_libcurl
  Referenced from: /usr/local/lib/python3.4/site-packages/pysam/libchtslib.so
  Expected in: flat namespace
 in /usr/local/lib/python3.4/site-packages/pysam/libchtslib.so
```

**Comment by AndreasHeger** [comment]
Thanks for the quick feed-back.

**Comment by AndreasHeger** [comment]
@leekaiinthesky , I wonder if you have the complete log? I'd be interested to see if there is some clue in the system configuration and the linker parameters.

**Comment by leekaiinthesky** [comment]
Sorry, I'm not sure what complete log means? Does `python -v` give what you want?

Here's the output when I do `import pysam` with `-v` in Python 2.7: [2.7_log.txt](https://github.com/pysam-developers/pysam/files/344170/2.7_log.txt), and here it is for Python 3.4: [3.4_log.txt](https://github.com/pysam-developers/pysam/files/344169/3.4_log.txt).

Let me know if that's not what you were looking for.

**Comment by AndreasHeger** [comment]
Thanks, apologies that I was not clear. Do you still have the output from when you installed pysam with 
pip install?

**Comment by jmarshall** [comment]
A missing `hfile_plugin_init_libcurl()` would mean that _plugin.c_ was compiled with `HAVE_LIBCURL` set and `ENABLE_PLUGINS` not set, but _hfile_libcurl.o_ did not make it into the library.

If this is limited to OS X, it is likely because the [_pysam: libcurl has not been found - disabled_ test](https://github.com/pysam-developers/pysam/blob/master/setup.py#L290-L292) does not account for `HAVE_COMMONCRYPTO`.  This might be a demonstration of how misguided this sort of code is, but that would be a digression :smile:

**Comment by leekaiinthesky** [comment]
Here's the verbose output when installing:

$ pip3 --verbose install pysam >pip_install_pysam_2.7.txt
[pip_install_pysam_2.7.txt](https://github.com/pysam-developers/pysam/files/344304/pip_install_pysam_2.7.txt)

$ pip3 --verbose install pysam >pip_install_pysam_3.4.txt
[pip_install_pysam_3.4.txt](https://github.com/pysam-developers/pysam/files/344305/pip_install_pysam_3.4.txt)

**Comment by jmarshall** [comment]
@AndreasHeger Please try pulling jmarshall/htslib@155e1aacac741febd6fd722cf7e3660da38d6a9b into Pysam's copy of HTSlib.  If this `print-config` target solves your problems, it will make its way into HTSlib proper.

With this, you will be able to parse the output of `cd …/htslib; make print-config` and use especially `LIBHTS_OBJS` and `LIBS` to more robustly replace [the _hfile_libcurl.c_ and `curl`/`crypto` selecting code](https://github.com/pysam-developers/pysam/blob/master/setup.py#L263-L304) for builtin htslib.

**Comment by odoland** [comment]
On Mac OS X 10.11.2, installed with pip yesterday (version 0.9.1.3)

Also same error, 

Checking for required python modules (/usr/bin/python)...
/2.7/site-packages/pysam/libchtslib.so, 2): Symbol not found: _hfile_plugin_init_libcurl
  Referenced from: /Library/Python/2.7/site-packages/pysam/libchtslib.so
  Expected in: flat namespace
 in /Library/Python/2.7/site-packages/pysam/libchtslib.so


@jmarshall Sorry for asking this, as I am a newbie, but how can I pull jmarshall/htslib@ 155e1aa into Pysam's copy of HTSlib?

**Comment by leekaiinthesky** [comment]
@odoland I believe that note was for the pysam developers. If you're ok just using a prior version of pysam, you could install a previous version for now, e.g.: `pip install pysam==0.8.4`

**Comment by AndreasHeger** [comment]
@jmarshall , many thanks, that will be very helpful. I am currently swamped with work, will try to do this weekend.

Best wishes,
Andreas

**Comment by ernesto-up** [comment]
I also had the import error :

ImportError: dlopen(/Users/Ernesto/Library/Python/2.7/lib/python/site-packages/pysam-0.9.1.3-py2.7-macosx-10.10-intel.egg/pysam/libchtslib.so, 2): Symbol not found: _hfile_plugin_init_libcurl
  Referenced from: /Users/Ernesto/Library/Python/2.7/lib/python/site-packages/pysam-0.9.1.3-py2.7-macosx-10.10-intel.egg/pysam/libchtslib.so
  Expected in: flat namespace
 in /Users/Ernesto/Library/Python/2.7/lib/python/site-packages/pysam-0.9.1.3-py2.7-macosx-10.10-intel.egg/pysam/libchtslib.so

Details:
macosx-10.10.5
python2.7
pysam 9.1.3, Installed by downloading the compressed zip file and running: 
>python setup.py build
>python setup.py install
 I'm a newbie so sorry if my explanation of the error is not clear.

**Comment by odoland** [comment]
@ernesto-up 
`pip install pysam==0.8.4`

As leekaiinthesky mentioned,  install the older version.

**Comment by csmit-rm** [comment]
Same issue. Also tried the older version and just got a different error:
pysam.csamfile.Samfile has the wrong size

Any insights = much appreciated

**Comment by kyleabeauchamp** [comment]
I recommend trying to install with miniconda + bioconda:

https://bioconda.github.io/

http://conda.pydata.org/miniconda.html

`conda install -c bioconda -c R pysam`.

**Comment by ernesto-up** [comment]
For me installing an older version solved it.

**Comment by leekaiinthesky** [comment]
@csmit-rm That doesn't look like an installation error. Can you please give the full error output and what you did that caused it? It looks like maybe you installed the prior version of pysam successfully, and then you experienced an unrelated error afterward, but it's hard to tell without more detail.

**Comment by csmit-rm** [comment]
I solved it by uninstalling pysam and pysamstats

export HTSLIB_LIBRARY_DIR=/usr/local/lib
export HTSLIB_INCLUDE_DIR=/usr/local/include

reinstalling pysam and pysamstats

**Comment by widdowquinn** [comment]
I cannot replicate the fix suggested by @csmit-rm above (OSX 10.11.5, Python2.7, virtualenv), but reversion to `pysam` v0.8.4 does work for me.

**Comment by martinbc** [comment]
Version 0.9.1 works on OSX 10.11.5, python3 and virtualenv.

**Comment by AndreasHeger** [comment]
Hi all, I have implemented @jmarshall suggestion and pushed a new release (0.9.1.4) out. Please let me know if this solves any of the installation issues in this thread.

**Comment by leekaiinthesky** [comment]
Yes. 0.9.1.4 resolves this issue (for both Python 2.7 and 3.4) on OS X. Thanks @AndreasHeger and @jmarshall!

**Comment by edwwlui** [comment]
in my case restarting the terminal solved the problem
